staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/complete.go

```diff
@@ -66,17 +66,29 @@ func validateNestedValueValidationCompleteness(v *NestedValueValidation, s *Stru
 
 	allErrs := field.ErrorList{}
 
-	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath)...)
-	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"))...)
+	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath, opts)...)
+	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"), opts)...)
+
+	var additionalPropertiesSchema *Structural
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		additionalPropertiesSchema = s.AdditionalProperties.Structural
+	}
+
 	for k, vFld := range v.Properties {
 		if sFld, ok := s.Properties[k]; !ok {
-			allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			if additionalPropertiesSchema == nil || !opts.AllowValidationPropertiesWithAdditionalProperties {
+				allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			} else {
```

Comment:
not sure this if-else is correct. Don't we have to follow both paths?

Also note my question elsewhere whether `additionalProperties` and `properties` can exist in parallel. If not, we have a unique path in the structural schema for every specified field in the value validation. Without that condition it's becoming exponential with the depth of nesting potentially.

And if we have that rule, then the else case can never happen.